### PR TITLE
Execute shell commands through /bin/sh

### LINKED ISSUE #134: Execute shell commands through a shell
Not using a shell leads to unexpected behaviour, e.g. without a shell to interpret the `~` in the argument, the command `!vim ~/foo` will edit the file `./~/foo`

### PR DESCRIPTION
**Pull request by georgebrock**
Resolves #134

### CODE REVIEW

File: lib/gitsh/commands/shell_command.rb [hunk 1]
```diff
@@ -22,11 +22,15 @@ def execute
...
+        [
+          '/bin/sh',
+          '-c',
+          [command, arg_values].flatten.join(' '),
+        ]
       end
 
       def arg_values
-        args.values(env)
+        args.values(env).map { |arg| arg.gsub(' ', '\ ') }
```
Comment (mike-burns) [inline lib/gitsh/commands/shell_command.rb]: Should we use `Shellwords` here?

Comment (georgebrock) [inline lib/gitsh/commands/shell_command.rb]: Good call.

**Comment by mike-burns** [comment]
After `Shellwords`: LGTM.

**Comment by georgebrock** [comment]
I'm having a little trouble with `Shellwords`: some things (like spaces in the middle of arguments) ought to be escaped, but other things (like globs) ought not be escaped. `Shellwords` will escape all of them.

A few examples:

* `!echo Hello world` should execute `/bin/sh -c "echo Hello world"`
* `!echo "Hello world"` should execute `/bin/sh -c "echo Hello\ world"`
* `!ls *` should execute `/bin/sh -c "ls *"`

**Comment by mike-burns** [comment]
I don't see a solution using `Shellwords` directly.

Interestingly, `shellwords.rb` is incredibly short and mostly consists of:

```ruby
def shellescape(str)
  str = str.to_s
  return "''" if str.empty?
  str = str.dup
  str.gsub!(/([^A-Za-z0-9_\-.,:\/@\n])/, "\\\\\\1")
  str.gsub!(/\n/, "'\n'")
  str
end
```

(Read the file itself for more details.)

Seems smart to use a whitelist like they do instead of a blacklist. Maybe this regexp: `[^A-Za-z0-9_\-.,:\/@\n*?\[\!]` -- pulling from [glob(7)](http://man.openbsd.org/glob.7).

**Comment by georgebrock** [comment]
@mike-burns All great ideas! I added custom escaping, and tried to make it explicit what was based on `Shellwords` and what was added to avoid escaping globbing patterns. Let me know what you think.

File: lib/gitsh/commands/shell_command.rb [hunk 2]
```diff
@@ -1,8 +1,14 @@
+require 'shellwords'
```
Comment (mike-burns) [inline lib/gitsh/commands/shell_command.rb]: Is this used?

**Comment by mike-burns** [comment]
Looks great!

**Pull request merged**